sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -2142,22 +2142,118 @@ case class OctetLength(child: Expression)
   """,
   since = "1.5.0",
   group = "string_funcs")
-case class Levenshtein(left: Expression, right: Expression) extends BinaryExpression
-    with ImplicitCastInputTypes with NullIntolerant {
+case class Levenshtein(
+    left: Expression,
+    right: Expression,
+    threshold: Option[Expression] = None)
+  extends Expression
+  with ImplicitCastInputTypes
+  with NullIntolerant{
 
-  override def inputTypes: Seq[AbstractDataType] = Seq(StringType, StringType)
+  def this(left: Expression, right: Expression, threshold: Expression) =
+    this(left, right, Option(threshold))
+
+  def this(left: Expression, right: Expression) = this(left, right, None)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.length > 3 || children.length < 2) {
+      throw QueryCompilationErrors.wrongNumArgsError(
+        toSQLId(prettyName), Seq(2, 3), children.length)
+    } else {
+      super.checkInputDataTypes()
+    }
+  }
+
+  override def inputTypes: Seq[AbstractDataType] = if (threshold.isDefined) {
+    Seq(StringType, StringType, IntegerType)
+  } else {
+    Seq(StringType, StringType)
+  }
+
+  override def children: Seq[Expression] = if (threshold.isDefined) {
+    Seq(left, right, threshold.get)
+  } else {
+    Seq(left, right)
+  }
 
   override def dataType: DataType = IntegerType
-  protected override def nullSafeEval(leftValue: Any, rightValue: Any): Any =
-    leftValue.asInstanceOf[UTF8String].levenshteinDistance(rightValue.asInstanceOf[UTF8String])
 
-  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    nullSafeCodeGen(ctx, ev, (left, right) =>
-      s"${ev.value} = $left.levenshteinDistance($right);")
+  override protected def withNewChildrenInternal(newChildren: IndexedSeq[Expression]): Expression =
+    if (threshold.isDefined) {
+      copy(left = newChildren(0), right = newChildren(1), threshold = Some(newChildren(2)))
+    } else {
+      copy(left = newChildren(0), right = newChildren(1))
+    }
+
+  override def nullable: Boolean = children.exists(_.nullable)
+
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  override def eval(input: InternalRow): Any = {
+    val leftEval = left.eval(input)
+    if (leftEval == null) return null
+    val rightEval = right.eval(input)
+    if (rightEval == null) return null
+
+    val thresholdEval = threshold.map(_.eval(input))
+    if (thresholdEval.isDefined) {
+      leftEval.asInstanceOf[UTF8String].levenshteinDistance(
+        rightEval.asInstanceOf[UTF8String], thresholdEval.get.asInstanceOf[Int])
+    } else {
+      leftEval.asInstanceOf[UTF8String].levenshteinDistance(rightEval.asInstanceOf[UTF8String])
+    }
   }
 
-  override protected def withNewChildrenInternal(
-    newLeft: Expression, newRight: Expression): Levenshtein = copy(left = newLeft, right = newRight)
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
```

Comment:
There are many branches here to check `if (threshold.isDefined)`. 

If we split it into 
```scala
if (threshold.isDefined) { // or threshold match ...
doGenCodeWithThreshold(...)
} else {
doGenCodeWithoutThreshold(...)
}
```
, will it look clearer?

Reply:
Done